src/EditorFeatures/CSharp/ConvertNamespace/ConvertNamespaceCommandHandler.cs

```diff
@@ -0,0 +1,168 @@
+
+            // No matter if we succeeded or not, insert the semicolon.  This way, when we convert, the user can still
+            // hit ctrl-z to get back to the code with just the semicolon inserted.
+            nextCommandHandler();
+
+            // If we weren't on a block namespace (or couldn't convert it for some reason), then bail out after
+            // inserting the semicolon.
+            if (convertedRoot == null)
+                return;
+
+            // Otherwise, make a transaction for the edit and replace the buffer with the final text.
+            using var transaction = CaretPreservingEditTransaction.TryCreate(
+                this.DisplayName, args.TextView, _textUndoHistoryRegistry, _editorOperationsFactoryService);
+
+            var edit = args.SubjectBuffer.CreateEdit(EditOptions.DefaultMinimalChange, reiteratedVersionNumber: null, editTag: null);
+            edit.Replace(new Span(0, args.SubjectBuffer.CurrentSnapshot.Length), convertedRoot.ToFullString());
+
+            edit.Apply();
+
+            // Attempt to place the caret right after the semicolon of the file-scoped namespace.
+            var annotatedToken = convertedRoot.GetAnnotatedTokens(s_annotation).FirstOrDefault();
+            if (annotatedToken != default)
+                args.TextView.Caret.MoveTo(new SnapshotPoint(args.SubjectBuffer.CurrentSnapshot, annotatedToken.Span.End));
+
+            transaction?.Complete();
+        }
+
+        /// <summary>
+        /// Returns true if semicolon is typed after a namespace name that should be converted.
+        /// </summary>
+        private static CompilationUnitSyntax? ConvertNamespace(
+            TypeCharCommandArgs args,
+            CommandExecutionContext executionContext)
+        {
+            if (args.TypedChar != ';' || !args.TextView.Selection.IsEmpty)
+                return null;
+
+            var subjectBuffer = args.SubjectBuffer;
+            var caretOpt = args.TextView.GetCaretPoint(subjectBuffer);
+            if (!caretOpt.HasValue)
+                return null;
+
+            var caret = caretOpt.Value.Position;
+            var document = subjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+                return null;
+
+            var cancellationToken = executionContext.OperationContext.UserCancellationToken;
+            var root = (CompilationUnitSyntax)document.GetRequiredSyntaxRootSynchronously(cancellationToken);
+
+            // User has to be *after* an identifier token.
+            var token = root.FindToken(caret);
+            if (token.Kind() != SyntaxKind.IdentifierToken)
+                return null;
+
+            if (caret < token.Span.End ||
+                caret >= token.FullSpan.End)
+            {
+                return null;
+            }
+
+            var namespaceDecl = token.GetRequiredParent().GetAncestor<NamespaceDeclarationSyntax>();
```

Comment:
Can a token actually have a null parent? ie, is it possible this can crash in error scenarios?

Reply:
a token can have a null parent in the case that the token is `default`.  But we don't have that as we have an actual Identifier token.  In that case, it's a requirement htat it has a parent as part of the general invariants of the Roslyn syntax model.